### ISSUE TITLE
Accessibility: Text is unreadably small when viewport width is 320px

### ISSUE THREAD

**Issue by rezinak** [highcharts/highcharts#18524]
#### Expected behaviour
The text still needs to be readable even on small devices.
([WCAG Reflow Criteria](https://www.w3.org/TR/WCAG21/#reflow))

#### Actual behaviour
Content is unreadably small when the viewport width is set to 320 pixels.
Users with low vision who need to resize text will be prevented from accessing this content.

#### Live demo with steps to reproduce
https://www.idc.com/QA_test_section/accessibility-remediation/highcharts-demo1
Set the viewport width to 320 px.

#### Product version
Highcharts JS v10.3.3

#### Affected browser(s)
Chrome and Mozilla, haven't looked into the others

**Comment by oysteinmoseng** [comment]
This is a usability problem for sure, and we have plans to introduce controls for the end user to increase font size on demand. I am not sure we can classify it as a bug, however, as the text scales as expected (relative to the other elements) - at least on my end.

You can override the relative size of the fonts by setting responsive rules: https://jsfiddle.net/oysteinmoseng/1gm5hqyc/ 

For now I am marking this as a feature request, but let me know if I have misunderstood and there is behavior here that is clearly buggy.

**Comment by TorsteinHonsi** [comment]
@oysteinmoseng I'm not sure I understand this - is it about the x axis labels? They are 11px by default regardless of the viewport width... The only difference is that it is rotated 45 degrees. Does the WCAG require the font size to be _greater_ at small screens?

It would be interesting to learn more here because we are considering increasing the default axis label font size for HC v11.

**Comment by oysteinmoseng** [comment]
WCAG does not give a minimum font size, but generally requires text to be resizable. The smallest font size used by default is in credits, which is 9px, not sure if this question was specific to axis labels (@rezinak?).

**Comment by rezinak** [comment]
> WCAG does not give a minimum font size, but generally requires text to be resizable. 

Yep, that's another problem. People with low vision or dyslexia may override text spacing to enable readability or increased reading speed. You should ensure that when users override text spacing via style sheet or another browser setting, there is no loss of content or functionality - text must not be cut off or missing. 

> The smallest font size used by default is in credits, which is 9px, not sure if this question was specific to axis labels

You have 9px for the axis labels and 11px for other texts. There isn't an official WCAG rule for minimal required font size, but it is generally agreed upon that 16px for body text is a good starting point. Some text can be smaller, but 9px is not readable for many people and 11px is borderline unreadable. Especially considering the impossibility to override text properties manually. I'll confirm with our accessibility consultant what the font-size values in this case should be exactly, and get back to you.

**Comment by TorsteinHonsi** [comment]
Thanks for elaborating @rezinak !

> You have 9px for the axis labels and 11px for other texts

Just for the records, the [default axis label font size is 11px](https://api.highcharts.com/highcharts/xAxis.labels.style.fontSize), so if you're seeing 9px it must be part of some non-default configuration.

**Comment by TorsteinHonsi** [comment]
As said above, we're considering increasing the default font size for v11. 16px is probably not practical because it comes with its own cost, which is that more text will be truncated because it requires more space, and we need a certain number of axis labels in order to convey the scale and categories. So it will be a trade off. 

> no loss of content or functionality - text must not be cut off or missing

This must be up to the author of the chart (the one who implements it), not up to the library. If someone for example creates a chart with too much text in the axis categories, we have no option but to truncate it or leave it out, otherwise it would overlap.

On another note, @oysteinmoseng and @rezinak , would it help make our library more accessible if we changed all the default font sizes to be based on `rem` units rather than `px`? Then it would be easier for the web page authors to override all the font sizes in one setting?

**Comment by oysteinmoseng** [comment]
@TorsteinHonsi Yes, `rem` (or maybe `em`) units should definitely be used - I think Nancy is already on this, but probably smart to verify!

Regarding text spacing and other text alterations through CSS, this should work already, but some tools apply these styles after the chart has loaded, in which case the Highcharts text metrics end up being wrong, and text may be cut off. See for example https://github.com/highcharts/highcharts/issues/15151

My thought on the font size issue in general is that because Highcharts text will often be smaller than surrounding content it would be good to have functionality for the end user to override font size specifically on the chart. This would also be good in cases where the Highcharts author has used `px` units to set non-overridable specific font sizes. We have talked about introducing an accessibility menu/menuitem, and this could fit in nicely there with minimal code on our side. Any opinions on this approach?

**Comment by TorsteinHonsi** [comment]
@oysteinmoseng 

Alerting font sizes after the chart has loaded as you say, has its problems because the layout is computed in JavaScript, and simply changing the CSS doesn't update the chart. 

I think we have the tools necessary to compute correct bounding boxes based on `rem` or `em` units, so it is up to us to decide. Pixels were chosen back in the day because it is simple to translate pixels to, well, pixels. But since we introduced styled mode we created ways to measure ems which will probably work on `rem` too. See https://www.highcharts.com/samples/highcharts/css/em for the styled mode variant. 

Nancy hasn't been assigned to work on this. If we do it, it's more of a developer job to replace all the pixel font sizes with `rem` or `em` and set up demos for a global or container-specific override.

**Comment by TorsteinHonsi** [comment]
Thanks for commenting @iamtheddrman . We will investigate if there is a good way to watch for style changes that will affect the layout. 

A small note though - Since Highcharts v11, text elements have relative font size, so the font _will_ respond to changes in the page CSS. What will not respond however is the chart layout. The consequence may be text labels overlapping.

**Comment by Xillez** [comment]
Regarding detecting CSS injections, see this comment => https://github.com/highcharts/highcharts/issues/15151#issuecomment-2037493784
